Add scripted and declarative Pipeline examples (#401)
jenkinsci/lockable-resources-plugin: src/doc/examples/lock-nodes.md

# Examples

## Node depended resources

> Create resources that are node depended is in many cases very helpful. That means a job must pick a target node where the requested resource is available.

Reviewer: Sentence per line and some minor phrasing proposals
```suggestion
Locking a resource that depends on a specific node can be very helpful in many cases.
That means a job must pick a target node that has the requested resource available.
```

```groovy
// allocate node
node('some-build-node') {
  // Lock resource named *whatever-resource-some-build-node*
  lock("whatever-resource-${env.NODE_NAME}") {

>     echo "I am no node ${env.NODE_NAME} and locked resource ${env.LOCKED_RESOURCE}"

Reviewer: ```suggestion
    echo "Running on node ${env.NODE_NAME} with locked resource ${env.LOCKED_RESOURCE}"
```

  }
}
```

> But much more useful is lock node first.

Reviewer: This example does not acquire the lock on `whatever-resource-${env.NODE_NAME}"`.  Shouldn't it acquire that lock in order to be consistent with the earlier example?
```suggestion
But it is much more useful to lock the node before locking the node-specific resource.
```

```groovy
// Lock resource named *some-build-node*
lock('some-build-node') {
  // allocate node
  node(env.LOCKED_RESOURCE) {

>     echo "I am no node ${env.NODE_NAME} and locked resource ${env.LOCKED_RESOURCE}"

Reviewer: ```suggestion
    echo "Running on node ${env.NODE_NAME} with locked resource ${env.LOCKED_RESOURCE}"
```

  }
}
```

> Let explain in more complex use case.

Reviewer: ```suggestion
Here is a more complex use case.
```

*Request:*

> Our job tests server-client integration. That means we need 2 nodes (1 server and 1 client).
> On every node must be test sources up-to-date.
> Tests are running only on client side.

Reviewer: ```suggestion
Our tests check operations between a client and a server.
The tests need 2 nodes, a server node and a client node.
Test sources must be updated on every node.
Tests run only on the client node.
```

*Solution:*
Create 2 nodes:

- node-server
- node-client

and execute it parallel like this:

```groovy
Map stages = [:];
stages['server'] = {
  node('node-server') {
    prepareTests()
    startServer()
  }
}
stages['client'] = {
  node('node-client') {
    prepareTests()
    startClientTest()
  }
}

> // execute all prepare stages synchronous

Reviewer: I though that parallel would start the two nodes at roughly the same time and allow them to both proceed .  Wouldn't that be "concurrently" rather than "synchronously"?
```suggestion
// execute all stages concurrently
```

parallel stages

// Prepare tests on node
void prepareTests() {

>   checkout([$class: 'GitSCM',
>             branches: [[name: 'master']]
>            ])

Reviewer: I think that we don't want to encode the SCM type or the branch name here because that is provided by the multibranch project where we are running this job.  By using a multibranch job, that allows me to test changes to the environment without altering the source code of the test.  It will use the branch that provided the Jenkinsfile
```suggestion
  checkout scm
```

}
// Start server
void startServer() {

>     echo 'Server will be started in few seconds'

Reviewer: ```suggestion
    echo 'Starting server'
```

    sh 'mvn clean hpi:run'
    echo 'Server is done'
}
// Start client
void startClientTest() {
  sleep 20
  sh 'mvn clean verify'
}
```

It looks pretty fine and easy, but !!!.

> Executing all steps parallel might leads to timing issues, because checkout on server-node might takes much longer then on client-node. This is serious issue because the client starts before the server and can not connect to server.

Reviewer: ```suggestion
Executing all steps in parallel will lead to timing issues.
Allocation of the server node might be delayed because a server node is not available.
Checkout on the server node might take much longer than on the client node.
This is serious issue because the client starts before the server and can not connect to server.
```

The solution is to synchronized parallel stages like this.

```groovy
Map prepareStages = [:];
prepareStages['server'] = {
  node('node-server') {
    prepareTests()
  }
}
prepareStages['client'] = {
  node('node-client') {
    startServer()
  }
}

> // execute all prepare stages synchronous

Reviewer: ```suggestion
// execute all prepare stages concurrently
```

parallel prepareStages

Map testStages = [:]
testStages['server'] = {
  node('node-server') {
    prepareTests();
  }
}
testStages['client'] = {
  node('node-client') {
    sleep 20
    startClientTest();
  }
}

// execute all test stages at the same time
testStages.failFast = true
parallel testStages

...
```

> Ok we solve the timing issue, but what is wrong here?

Reviewer: ```suggestion
The timing issue looks like it is solved, but something is wrong.
```

> When the step *parallel prepareStages* is done then are on both nodes executors free. At this moment
> it might happen, that some other job allocate one of the nodes. This will leads to more side effects, like:

Reviewer: ```suggestion
When the *parallel prepareStages* step finishes, both node executors are freed.
Another job might allocate one of the nodes. 
This will leads to more side effects, like:
```

> - no body can grant, that currently checked out workspace will be untouched
> - no body can grant how long will be the node allocated
> - ... and many others

Reviewer: ```suggestion
- no assurance that the node is still available
- no assurance that the workspace is unmodified
- ... and many others
```

> Possible solution weill be to locked the nodes.

Reviewer: ```suggestion
Instead, we lock both nodes with a single call to `lock`.
```

> Therefore you need to create 2 resources

Reviewer: ```suggestion
Create two resources:
```

name           | Labels |
---------------|--------|
nodes-server-1 | server-node |
nodes-client-1 | client-node  |

```groovy
lock(variable: 'locked_resources',
     extra: [
      [label: 'server-node', quantity: 1],
      [label: 'client-node', quantity: 1]
    ) {
  final String serverNodeName = env.LOCKED_RESOURCE0;
  final String clientNodeName = env.LOCKED_RESOURCE1;
  Map prepareStages = [:];
  prepareStages['server'] = {
    node(serverNodeName) {
      prepareTests()
    }
  }
  prepareStages['client'] = {
    node(clientNodeName) {
      startServer()
    }
  }
  // execute all prepare stages synchronous
  parallel prepareStages

  Map testStages = [:]
  testStages['server'] = {
    node(serverNodeName) {
      prepareTests();
    }
  }
  testStages['client'] = {
    node(clientNodeName) {
      sleep 20
      startClientTest();
    }
  }

  // execute all test stages at the same time
  testStages.failFast = true
  parallel testStages

}

...
```

> Keep in mind, that *lock()* helps only when you use it in other jobs too.

Reviewer: ```suggestion
Keep in mind, that `lock()` only helps when locks are consistently requested for resources.
```
